ui/org.eclipse.pde.ui.tests/src/org/eclipse/pde/ui/tests/target/IUBundleContainerTests.java

```diff
@@ -327,9 +377,58 @@ public void testExternalModelManagerPreferences() throws Exception {
 	}
 
 	/**
-	 * Creates an IU bundle container with the specified IUs from the test repository.
+	 * Creates an in-memory artifact repository over {@code test/sites/site.a.b}
 	 *
-	 * @param unitIds identifiers of IU's to add to the container
+	 * @return in-memory artifact repository
+	 */
+	protected VirtualArtifactRepository createVirtualArtifactRepository() throws Exception {
+		File repoFolder = new File(getURI("/tests/sites/site.a.b"));
+		// DirectoryBundleContainer expects features to be unpacked
+		File features = new File(repoFolder, "features");
+		for (File jarFile : features.listFiles()) {
+			if (jarFile.isDirectory()) {
+				continue;
+			}
+			String jarFileName = jarFile.getName();
+			String jarFileBaseName = jarFileName.substring(0, jarFileName.length() - ".jar".length());
+
+			File outDir = new File(features, jarFileBaseName);
+			outDir.mkdir();
+			outDir.deleteOnExit();
+
+			try (JarFile jar = new JarFile(jarFile)) {
+				jar.stream().forEach(jarEntry -> {
+					File outFile = new File(outDir, jarEntry.getName());
+					outFile.deleteOnExit();
+					if (outFile.isDirectory()) {
+						outFile.mkdir();
+						return;
+					}
+					try (InputStream is = jar.getInputStream(jarEntry)) {
+						try (OutputStream os = new FileOutputStream(outFile)) {
+							is.transferTo(os);
+						}
+					} catch (IOException e) {
+						e.printStackTrace();
+						fail(e.getMessage());
```

Comment:
Doesn't fail provide an overload, where you can pass an exception?

Reply:
It only does in [JUnit5](https://junit.org/junit5/docs/5.0.1/api/org/junit/jupiter/api/Assertions.html#fail-java.lang.Throwable-), not in [JUnit4](https://junit.org/junit4/javadoc/latest/org/junit/Assert.html#fail(java.lang.String)).

The catch-block is only necessary because of the lambda-statement. Perhaps I can rewrite this part to use a simple for-each loop... Otherwise it's probably better to simply wrap the IOException in an unchecked exception. Just so that the stack trace is preserved, in case this operation _does_ fail.

Reply:
Also obsolete